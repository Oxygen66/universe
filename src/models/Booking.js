import database from '../../database';

class Booking {
  #db = database;

  async getAllByFlightId(flightId) {
    return this.#db('booking')
      .where('flight_id', flightId);
  }
}

export default Booking;
