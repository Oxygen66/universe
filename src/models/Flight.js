import crypto from 'crypto';
import database from '../../database';

class Flight {
  #db = database;

  async createFlight(launchSiteId, landingSiteId, departureAt, seatCount) {
    return this.#db('flight')
      .insert({
        launching_site: launchSiteId,
        landing_site: landingSiteId,
        departure_at: departureAt,
        seat_count: seatCount,
        code: crypto.randomBytes(16).toString('hex'),
      })
      .returning(['id', 'code', 'departure_at', 'seat_count', 'landing_site', 'launching_site']);
  }
}

export default Flight;
