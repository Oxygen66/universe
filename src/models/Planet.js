import database from '../../database';

class Planet {
  #db = database;

  async getAll() {
    return this.#db('planet');
  }

  async getBySpaceCenterId(id) {
    return this.#db('planet')
      .select('planet.*')
      .innerJoin(
        'space_center',
        'space_center.planet_id',
        'planet.id',
      )
      .where('space_center.id', id)
      .first();
  }
}

export default Planet;
