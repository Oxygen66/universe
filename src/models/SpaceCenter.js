import database from '../../database';

class SpaceCenter {
  #db = database;

  async getByPlanetIdAndLimit(planetId, limit) {
    return this.#db('space_center')
      .where('planet_id', planetId)
      .limit(limit);
  }

  async getById(id) {
    return this.#db('space_center')
      .where({ id })
      .first();
  }

  async getByUid(uid) {
    return this.#db('space_center')
      .where({ uid })
      .first();
  }

  async getAllPaginated(page, pageSize) {
    return this.#db('space_center')
      .limit(pageSize)
      .offset((pageSize - 1) * page);
  }

  async countAll() {
    return this.#db('space_center').count({ total: 'id' });
  }
}

export default SpaceCenter;
