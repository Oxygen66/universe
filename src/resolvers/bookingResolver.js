import { UserInputError } from 'apollo-server-koa';

const bookingResolver = {
  Mutation: {
    bookFlight: async (
      _,
      { bookingInfo: { seatCount, flightId, email } },
      { dataSources: { bookingAPI, flightAPI } },
    ) => {
      if (seatCount < 1) {
        throw new UserInputError('seatCount should be greater than 0', {
          invalidArgs: 'seatCount',
        });
      }

      const flight = await flightAPI.getById(flightId);

      const bookings = await bookingAPI.getAllByFlightId(flightId);
      const remainingSeats = [...bookings]
        .map((booking) => booking.seat_count)
        .reduce((previousValue, currentValue) => previousValue - currentValue, flight.seat_count);

      if ((remainingSeats - seatCount) < 0) {
        throw new UserInputError('seatCount must be less than the number of remaining seats', {
          invalidArgs: 'seatCount',
        });
      }

      const [bookingResult] = await bookingAPI.createBooking(seatCount, flightId, email);
      return bookingResult;
    },
  },
  Query: {
    booking: (_, { id }, { dataSources: { bookingAPI } }) => bookingAPI.getById(id),
  },
  Booking: {
    id: (parent) => parent.id,
    seatCount: (parent) => parent.seat_count,
    email: (parent) => parent.email,
    flight: (parent, args, { dataSources: { flightAPI } }) => flightAPI.getById(parent.flight_id),
  },
};

export default bookingResolver;
