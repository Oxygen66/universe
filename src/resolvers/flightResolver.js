import { UserInputError } from 'apollo-server-koa';

const flightResolver = {
  Mutation: {
    scheduleFlight:
      async (
        _,
        {
          flightInfo: {
            launchSiteId,
            landingSiteId,
            departureAt,
            seatCount,
          },
        },
        { dataSources: { flightAPI } },
      ) => {
        if (landingSiteId === launchSiteId) {
          throw new UserInputError('landingSiteId and landingSiteId cannot be the same', {
            invalidArgs: [
              'launchSiteId', 'landingSiteId',
            ],
          });
        }
        const [flightResult] = await flightAPI
          .createFlight(launchSiteId, landingSiteId, departureAt, seatCount);
        return flightResult;
      },
  },
  Flight: {
    id: (parent) => parent.id,
    code: (parent) => parent.code,
    departureAt: (parent) => parent.departure_at,
    seatCount: (parent) => parent.seat_count,
    launchSite: (
      parent, args, { dataSources: { spaceCenterAPI } },
    ) => spaceCenterAPI.getById(parent.launching_site),
    landingSite: (
      parent, args, { dataSources: { spaceCenterAPI } },
    ) => spaceCenterAPI.getById(parent.landing_site),
    availableSeats: async (
      { id, seat_count: seatCount }, args, { dataSources: { bookingAPI } },
    ) => {
      const bookings = await bookingAPI.getAllByFlightId(id);
      return [...bookings]
        .map((booking) => booking.seat_count)
        .reduce((previousValue, currentValue) => previousValue - currentValue, seatCount);
    },
  },
};

export default flightResolver;
