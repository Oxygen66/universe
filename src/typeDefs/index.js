import types from './types';
import query from './query';
import mutation from './mutation';
import inputs from './inputs';

const typeDefs = [
  ...types,
  ...inputs,
  query,
  mutation,
];

export default typeDefs;
