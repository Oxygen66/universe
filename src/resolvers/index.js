import { merge } from 'lodash';
import planetResolver from './planetResolver';
import spaceCenterResolver from './spaceCenterResolver';
import paginationResolver from './paginationResolver';
import flightResolver from './flightResolver';

const resolvers = merge(
  planetResolver,
  spaceCenterResolver,
  paginationResolver,
  flightResolver,
);

export default resolvers;
