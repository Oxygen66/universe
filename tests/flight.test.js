import { ApolloServer, gql } from 'apollo-server-koa';
import { createTestClient } from 'apollo-server-testing';
import Planet from '../src/models/Planet';
import SpaceCenter from '../src/models/SpaceCenter';
import Flight from '../src/models/Flight';
import typeDefs from '../src/typeDefs';
import resolvers from '../src/resolvers';
import Booking from '../src/models/Booking';

const planetAPI = new Planet();
const spaceCenterAPI = new SpaceCenter();
const flightAPI = new Flight();
const bookingAPI = new Booking();

const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: () => ({
    planetAPI, spaceCenterAPI, flightAPI, bookingAPI,
  }),
});

const CREATE_FLIGHT = gql`
    mutation scheduleFlight($flight: ScheduleFlightInput!) {
        scheduleFlight(flightInfo: $flight) {
            id
            code
            launchSite {
                name
                planet {
                    name
                }
            }
            landingSite {
                name
                planet {
                    name
                }
            }
            availableSeats
            seatCount
            departureAt
        }
    }
`;

describe('Mutations', () => {
  describe('scheduleFlight', () => {
    const fakeFlight = {
      id: 1,
      code: 'ddd99ffbca076ef4e188774eea6d4d05',
      departure_at: new Date('1970-01-01'),
      seat_count: 5,
      launching_site: 1,
      landing_site: 2,
    };

    beforeEach(() => {
      bookingAPI.getAllByFlightId = jest.fn(() => ([]));
      flightAPI.createFlight = jest.fn(() => [fakeFlight]);
      spaceCenterAPI.getById = jest.fn((id) => ({
        id,
        uid: `04c424d3-ff29-41b0-8d11-820c5e4c12b${id}`,
        name: `name${id}`,
        description: `description${id}`,
        planet_id: 1,
      }));
      planetAPI.getBySpaceCenterId = jest.fn(() => ({
        id: 1,
        name: 'test1',
        code: 'COD',
      }));
    });

    it('Create flight', async () => {
      const { mutate } = createTestClient(server);
      const flight = {
        launchSiteId: 1,
        landingSiteId: 2,
        departureAt: '1970-01-01',
        seatCount: 5,
      };
      const res = await mutate({
        mutation: CREATE_FLIGHT,
        variables: {
          flight,
        },
      });

      expect(res.errors).not.toBeTruthy();
      expect(res).toMatchSnapshot();
    });

    it('Create flight with same launchSiteId and landingSiteId throw error', async () => {
      const { mutate } = createTestClient(server);
      const res = await mutate({
        mutation: CREATE_FLIGHT,
        variables: {
          flight: {
            launchSiteId: 1,
            landingSiteId: 1,
            departureAt: '1970-01-01',
            seatCount: 5,
          },
        },
      });
      expect(res.errors).toBeTruthy();
      expect(res).toMatchSnapshot();
    });
  });
});
