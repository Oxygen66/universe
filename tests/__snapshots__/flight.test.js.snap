// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations scheduleFlight Create flight 1`] = `
Object {
  "data": Object {
    "scheduleFlight": Object {
      "availableSeats": 5,
      "code": "ddd99ffbca076ef4e188774eea6d4d05",
      "departureAt": "0",
      "id": "1",
      "landingSite": Object {
        "name": "name2",
        "planet": Object {
          "name": "test1",
        },
      },
      "launchSite": Object {
        "name": "name1",
        "planet": Object {
          "name": "test1",
        },
      },
      "seatCount": 5,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutations scheduleFlight Create flight with same launchSiteId and landingSiteId throw error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: landingSiteId and landingSiteId cannot be the same],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
